test/gurka/test_ramps_tc.cc

```diff
@@ -279,3 +279,148 @@ TEST(RampsTCs, test_tc_infer) {
+      {"FKC", {{"highway", "trunk_link"}, {"ref", "M2"}, {"oneway", "yes"}}},
+      // some tertiary road
+      {"KL", {{"highway", "tertiary"}}},
+  };
+
+  const auto layout = gurka::detail::map_to_coordinates(ascii_map, gridsize_metres);
+  auto map = gurka::buildtiles(layout, ways, {}, {}, "test/data/gurka_ramps_use_refs",
+                               {{"mjolnir.reclassify_links", "true"}});
+  baldr::GraphReader graph_reader(map.config.get_child("mjolnir"));
+
+  // check that ramp classification from M1 to M2 wasn't confused by the tertiary road
+  check_edge_classification(graph_reader, layout, "B", "J", baldr::RoadClass::kTrunk);
+  check_edge_classification(graph_reader, layout, "J", "F", baldr::RoadClass::kTrunk);
+
+  // check that ramp classification from M2 to M1 wasn't confused by the tertiary road
+  check_edge_classification(graph_reader, layout, "F", "K", baldr::RoadClass::kTrunk);
+  check_edge_classification(graph_reader, layout, "K", "C", baldr::RoadClass::kTrunk);
+}
+
+TEST(LinkReclassification, test_hierarchical_reclass) {
+  constexpr double gridsize_metres = 10;
+
+  const std::string ascii_map = R"(
+  A--------B-------------------------------------C
+            \                             D
+             \                            |
+              G-------------H-------------E
+              |             |             |
+          I---J---K         |             F
+                        L---M---N
+  )";
+
+  const gurka::ways ways = {
+      {"ABC", {{"highway", "motorway"}}},
+      {"DEF", {{"highway", "trunk"}}},
+      {"BGHE", {{"highway", "motorway_link"}, {"oneway", "yes"}}},
+      {"GJ", {{"highway", "motorway_link"}, {"oneway", "yes"}}},
+      {"HM", {{"highway", "motorway_link"}, {"oneway", "yes"}}},
+      {"IJK", {{"highway", "secondary"}}},
+      {"LMN", {{"highway", "tertiary"}}},
+  };
+
+  const auto layout = gurka::detail::map_to_coordinates(ascii_map, gridsize_metres);
+  auto map = gurka::buildtiles(layout, ways, {}, {}, "test/data/gurka_ramps_hierarchical_reclass",
+                               {{"mjolnir.reclassify_links", "true"}});
+  baldr::GraphReader graph_reader(map.config.get_child("mjolnir"));
+
+  // BG leads to 3 roads, make sure that the best classification was chosen
+  check_edge_classification(graph_reader, layout, "B", "G", baldr::RoadClass::kTrunk);
+  // GJ leads to only one secondary road
+  check_edge_classification(graph_reader, layout, "G", "J", baldr::RoadClass::kSecondary);
+  // GH lead to 2 road, make sure that the best classification was chosen
+  check_edge_classification(graph_reader, layout, "G", "H", baldr::RoadClass::kTrunk);
+  // HM leads to only one tertiary road
+  check_edge_classification(graph_reader, layout, "H", "M", baldr::RoadClass::kTertiary);
+  // HE leads to only one trunk road
+  check_edge_classification(graph_reader, layout, "H", "E", baldr::RoadClass::kTrunk);
+}
+
+TEST(LinkReclassification, test_acyclic_graph) {
```

Comment:
check that we correctly handle cases where graph nodes have several inbound edges and several outbound edges